File: VCL/CBuilder/GattConnections.cpp
//---------------------------------------------------------------------------

#pragma hdrstop

#include "GattConnections.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)

void __fastcall TGattConnections::ClientConnect(TObject* Sender,
	const int Error)
{
	if (Error != WCL_E_SUCCESS)
		Trace("Connection error: " + IntToHex(Error, 8));
	else
	{
		Trace("Client connected");

		TwclGattServices Services;
		int Res = FClient->ReadServices(goNone, Services);
		if (Res != WCL_E_SUCCESS)
			Trace("Failed to read services: " + IntToHex(Res, 8));
		else
			Trace("Services have been read with success");
	}
}

__fastcall TGattConnections::TGattConnections(TwclBluetoothRadio* const Radio,
	const __int64 Address, TListBox* const Log) : TwclThread()
{
	FAddress = Address;
	FRadio = Radio;
	FLog = Log;
}

bool __fastcall TGattConnections::OnInitialize()
{
	FClient = new TwclGattClient(NULL);
	FClient->OnConnect = ClientConnect;
	FClient->Address = FAddress;
	int Res = FClient->Connect(FRadio);
	if (Res != WCL_E_SUCCESS)
	{
		Trace("Start connection failed: " + IntToHex(Res, 8));
		FClient->Free();
		FClient = NULL;
		return false;
	}
	return true;
}

void __fastcall TGattConnections::OnSynchronize(const void* Param)
{
	FLog->Items->Add(*((String*)Param));
}

void __fastcall TGattConnections::OnTerminate()
{
	if (FClient != NULL)
	{
		FClient->Disconnect();
		FClient->Free();
		FClient = NULL;
	}
}

void __fastcall TGattConnections::Trace(String Msg)
{
	String Str = IntToHex(FAddress, 12) + ": " + Msg;
	Synchronize((void*)&Str);
}

File: VCL/Delphi/GattConnections.pas
unit GattConnections;

interface

uses
  wclThread, wclBluetooth, StdCtrls;

type
  TGattConnections = class(TwclThread)
  private
    FAddress: Int64;
    FClient: TwclGattClient;
    FRadio: TwclBluetoothRadio;
    FLog: TListBox;

    procedure Trace(const Msg: string);

    procedure ClientConnect(Sender: TObject; const Error: Integer);

  protected
    function OnInitialize: Boolean; override;
    procedure OnTerminate; override;
    procedure OnSynchronize(const Param: Pointer); override;

  public
    constructor Create(const Radio: TwclBluetoothRadio; const Address: Int64;
      const Log: TListBox); reintroduce;

    property Address: Int64 read FAddress;
  end;

implementation

uses
  SysUtils, wclErrors;

{ TGattConnections }

procedure TGattConnections.ClientConnect(Sender: TObject; const Error: Integer);
var
  Services: TwclGattServices;
  Res: Integer;
begin
  if Error <> WCL_E_SUCCESS then
    Trace('Connection error: ' + IntToHex(Error, 8))

  else begin
    Trace('Client connected');

    Res := FClient.ReadServices(goNone, Services);
    if Res <> WCL_E_SUCCESS then
      Trace('Failed to read services: ' + IntToHex(Res, 8))
    else
      Trace('Services have been read with success');
  end;
end;

constructor TGattConnections.Create(const Radio: TwclBluetoothRadio;
  const Address: Int64; const Log: TListBox);
begin
  inherited Create;

  FAddress := Address;
  FRadio := Radio;
  FLog := Log;
end;

function TGattConnections.OnInitialize: Boolean;
var
  Res: Integer;
begin
  FClient := TwclGattClient.Create(nil);
  FClient.OnConnect := ClientConnect;
  FClient.Address := FAddress;
  Res := FClient.Connect(FRadio);
  if Res <> WCL_E_SUCCESS then begin
    Trace('Start connection failed: ' + IntToHex(Res, 8));
    FClient.Free;
    FClient := nil;
    Result := False;
  end else
    Result := True;
end;

procedure TGattConnections.OnSynchronize(const Param: Pointer);
begin
  FLog.Items.Add(string(Param));
end;

procedure TGattConnections.OnTerminate;
begin
  if FClient <> nil then begin
    FClient.Disconnect;
    FClient.Free;
    FClient := nil;
  end;
end;

procedure TGattConnections.Trace(const Msg: string);
var
  Str: string;
begin
  Str := IntToHex(FAddress, 12) + ': ' + Msg;
  Synchronize(Pointer(Str));
end;

end.
